Source/DafnyCore/Verifier/Translator.ExpressionWellformed.cs

```diff
@@ -712,28 +712,52 @@ private record ReadsCheckDelayer(ExpressionTranslator etran, Function selfCallsA
                 }
               }
               // check that the preconditions for the call hold
-              foreach (AttributedExpression p in e.Function.Req) {
-                Expression precond = Substitute(p.E, e.Receiver, substMap, e.GetTypeArgumentSubstitutions());
-                var (errorMessage, successMessage) = CustomErrorMessage(p.Attributes);
-                foreach (var ss in TrSplitExpr(precond, etran, true, out var splitHappened)) {
-                  if (ss.IsChecked) {
-                    var tok = new NestedToken(GetToken(expr), ss.Tok);
-                    var desc = new PODesc.PreconditionSatisfied(errorMessage, successMessage);
-                    if (wfOptions.AssertKv != null) {
-                      // use the given assert attribute only
-                      builder.Add(Assert(tok, ss.E, desc, wfOptions.AssertKv));
-                    } else {
-                      builder.Add(AssertNS(tok, ss.E, desc));
+              // the check for .reads function must be translated explicitly: their declaration lacks
```

Comment:
This change is harder to review because it's part of a very large method, so it's not easy to find the method header and gather context about what this code is supposed to do.

I see this is part of a block to check the well-formedness of a call expression. Could we put that block in a separate method?

Also, could you see if you can come up with a method name that describes either the whole block between line 714 to 762 or one of the sub blocks, and if so put that in a separate method? Maybe something like `CheckReadsClauseSatisfied`

Reply:
Hmm. I wonder whether this change and the others you comment on below are an artifact of the PR not being up-to-date with `master`? Because they aren't changes made by this PR. There's some recent work @jtristan made to resolve a soundness issue that touches the same files, and that I had to resolve conflicts to merge with, and I think the changes you're pointing to are from that. Were you reviewing an intermediate commit, perhaps?